Enable TDX measurement to RTMR register

kenplusplus opened the pull request:
Intel Trust Domain Extensions(Intel TDX) refers to an Intel technology
that extends Virtual Machine Extensions(VMX) and Multi-Key Total Memory
Encryption(MK-TME) with a new kind of virtual machine guest called a
Trust Domain(TD)[1]. A TD runs in a CPU mode that protects the confidentiality
of its memory contents and its CPU state from any other software, including
the hosting Virtual Machine Monitor (VMM).

Trust Domain Virtual Firmware (TDVF) is required to provide TD services to
the TD guest OS.[2] Its reference code is available at https://github.com/tianocore/edk2-staging/tree/TDVF.

To support TD measurement/attestation, TDs provide 4 RTMR registers like
TPM/TPM2 PCR as below:
- RTMR[0] is for TDVF configuration
- RTMR[1] is for the TD OS loader and kernel
- RTMR[2] is for the OS application
- RTMR[3] is reserved for special usage only

This patch adds TD Measurement protocol support along with TPM/TPM2 protocol.

References:
[1] https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-whitepaper-v4.pdf
[2] https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-virtual-firmware-design-guide-rev-1.pdf

Signed-off-by: Lu Ken <ken.lu@intel.com>

vathpela commented on tdx.c:
```diff
@@ -0,0 +1,46 @@
+{
+	EFI_STATUS efi_status;
+	EFI_TCG2_EVENT *event;
+	efi_tdx_protocol_t *tdx;
+
+        efi_status = LibLocateProtocol(&EFI_TDX_GUID, (VOID **)&tdx);
+	if (EFI_ERROR(efi_status))
+		return EFI_SUCCESS;
+
+	UINTN event_size = sizeof(*event) - sizeof(event->Event) +
+		logsize;
+
+	event = AllocatePool(event_size);
+	if (!event) {
+		perror(L"Unable to allocate event structure\n");
+		return EFI_OUT_OF_RESOURCES;
+	}
+
+	event->Header.HeaderSize = sizeof(EFI_TCG2_EVENT_HEADER);
+	event->Header.HeaderVersion = 1;
+	event->Header.PCRIndex = pcr;
+	event->Header.EventType = type;
+	event->Size = event_size;
+	CopyMem(event->Event, (VOID *)log, logsize);
+	if (hash) {
+		/* TCG 2 systems will generate the appropriate hash
+			themselves if we pass PE_COFF_IMAGE.  In case that
+			fails we fall back to measuring without it.
+		*/
+		efi_status = tdx->hash_log_extend_event(tdx,
+			PE_COFF_IMAGE, buf, (UINT64) size, event);
+	}
+
+	if (!hash || EFI_ERROR(efi_status)) {
+		efi_status = tdx->hash_log_extend_event(tdx,
+			0, buf, (UINT64) size, event);
+	}
+	FreePool(event);
+	return efi_status;
+}
```
Please run `clang-format -i tdx.c` on this.  But also, is there a good reason not to put this in tpm.c/tpm.h?  It's fairly small and clearly related.

kenplusplus replied:
Thanks for your comments, I will clean up format. I will merge tdx.c -> tpm.c/ tdx.h -> tpm.h. I will create a new tcg2_log_event_raw() for both TPM and TDX. 
But unlike the exclusive of TPM and TPM2, the TDX maybe co-exist with TPM2 due to different technologies.
Will submit new patch later after validation.

vathpela reviewed (changes requested):
I think this should get merged into tpm.c/tpm.h instead of adding a new source file, and the formatting needs to be cleaned up, but otherwise this looks pretty good.  Thanks!

frozencemetery commented:
HI @kenplusplus, did you get a chance to create an updated patch?

kenplusplus commented:
> HI @kenplusplus, did you get a chance to create an updated patch?

@frozencemetery @vathpela Sorry for later update. @mxu9 is working on the new  EFI_TD_PROTOCOL in edk2 community at https://www.mail-archive.com/devel@edk2.groups.io/msg36487.html, which will be a little different with current TCG2 one.
I just updated the source code, and will resubmit here before end of this week.

kenplusplus commented:
> > HI @kenplusplus, did you get a chance to create an updated patch?
> 
> @frozencemetery @vathpela Sorry for later update. @mxu9 is working on the new EFI_TD_PROTOCOL in edk2 community at https://www.mail-archive.com/devel@edk2.groups.io/msg36487.html, which will be a little different with current TCG2 one.
> I just updated the source code, and will resubmit here before end of this week.

I re-submit the patches:
1. Update `struct efi_td_protocol` according to https://www.mail-archive.com/devel@edk2.groups.io/msg36488.html
2. Use "clang-format -i tdx.c include/tdx.h"

Since the TDX protocol is not exact match to TPM, so I still keep them in seperated header and source file.

frozencemetery commented on tdx.c:
```diff
@@ -0,0 +1,46 @@
+	}
+	FreePool(event);
+	return efi_status;
+}
```
Missing newline at end of file.

kenplusplus replied:
Thanks comment @frozencemetery, it was fixed at new commit https://github.com/rhboot/shim/commit/5f2abe426bb04fef2d2104f809260327af874c67, please help to review!

frozencemetery commented on include/tdx.h:
```diff
@@ -0,0 +1,99 @@
+
+typedef struct tdEFI_TD_EVENT {
+	//
+	// Total size of the event including the Size component, the header and the Event data.
```
Please check your line lengths.

kenplusplus replied:
Thanks comment @frozencemetery, it was fixed at new commit https://github.com/rhboot/shim/commit/5f2abe426bb04fef2d2104f809260327af874c67, please help to review!

frozencemetery commented on shim.h:
```diff
@@ -179,6 +179,7 @@
 #include "include/simple_file.h"
 #include "include/str.h"
 #include "include/tpm.h"
+#include "include/tdx.h"
```
I think vathpela's comment about this still stands.

kenplusplus replied:
@frozencemetery I hope we can separate the TDX and TPM's definition (header file) and implementation (C file), because:

- TDX([Trusted Domain Extension](https://software.intel.com/content/www/cn/zh/develop/articles/intel-trust-domain-extensions.html)) is designed for confidential VM. TPM is major designed for TPM device on the baremetal host.
- TDX use similiar structure on measurement like TPM, but not exact same according to recent update https://www.mail-archive.com/devel@edk2.groups.io/msg36488.html. So even put the TDX definition in TPM header file, the actual source code could not be reused much.
- Grub2 provide grub_file_verifier framework https://github.com/rhboot/grub2/blob/master/grub-core/commands/tpm.c#L80, you can register TPM, TDX, or other approach to achieve file verifier functionality. In today's shim, there is no such framework, so I have to call tdx_log_measurement() within tpm.c, which looks not good, but sound like there is no other good way.

kenplusplus commented:
@frozencemetery @vathpela I have re-submitted patch for:
1. Change EFI_TD_PROTOCOL => EFI_CC_MEASUREMENT_PROTOCOL, due to changes at EFI tianocore https://bugzilla.tianocore.org/show_bug.cgi?id=3625
2. Merge logics into tpm.c according to @vathpela comment, but still keep cc.h for protoctol declaration since EFI_CC_MEASUREMENT_PROTOCOL is not equal to TPM/TCG.

Could you please help for review and merge? Thanks!

frozencemetery commented:
Unrelated to all that, I noticed that you'd targeted this PR to 15.4.  I've moved it to main (where it would need to land), but it looks like there are conflicts now.

kenplusplus commented:
> Unrelated to all that, I noticed that you'd targeted this PR to 15.4. I've moved it to main (where it would need to land), but it looks like there are conflicts now.

@frozencemetery missed the conflict, just fixed it at commit https://github.com/rhboot/shim/pull/388/commits/d9036974495b786066ce9c0ff2f934f33ffdde33 and verified in my side.

frozencemetery commented:
That doesn't look right - did you push the right thing?  What you've got there is a merge commit.  We're looking for a rebase.

kenplusplus commented:
Sorry for later response, since wait for the review of similar grub's patch at https://www.mail-archive.com/grub-devel@gnu.org/msg34244.html. I rebased this patch to shim 15.5 and resubmitted PR at https://github.com/rhboot/shim/pull/479. Thanks!
